src/Stratis.Bitcoin/P2P/Peer/NetworkPeerListener.cs

```diff
@@ -1,78 +1,71 @@
 ﻿using System;
-using System.Collections.Generic;
-using System.Linq;
 using System.Threading;
-using NBitcoin.Protocol;
+using System.Threading.Tasks;
 using Stratis.Bitcoin.P2P.Protocol;
 using Stratis.Bitcoin.P2P.Protocol.Payloads;
+using Stratis.Bitcoin.Utilities;
 
 namespace Stratis.Bitcoin.P2P.Peer
 {
-    public class NetworkPeerListener : PollMessageListener<IncomingMessage>, IDisposable
+    /// <summary>
+    /// Message listener that waits until a specific payload is received and returns it to the caller.
+    /// </summary>
+    public class NetworkPeerListener : IMessageListener<IncomingMessage>, IDisposable
     {
-        public NetworkPeer Peer { get; private set; }
-        private IDisposable subscription;
-        private List<Func<IncomingMessage, bool>> predicates = new List<Func<IncomingMessage, bool>>();
+        /// <summary>Queue of unprocessed messages.</summary>
+        private readonly AsyncQueue<IncomingMessage> asyncQueue;
 
+        /// <summary>Connected network peer that we receive messages from.</summary>
+        private readonly NetworkPeer peer;
+
+        /// <summary>Subscription to the message producer of the connected peer.</summary>
+        private readonly IDisposable subscription;
+
+        /// <summary>
+        /// Initializes the instance of the object and subscribes to the peer's message producer.
+        /// </summary>
+        /// <param name="peer">Connected network peer that we receive messages from.</param>
         public NetworkPeerListener(NetworkPeer peer)
         {
+            this.asyncQueue = new AsyncQueue<IncomingMessage>();
```

Comment:
I am considering we should rename `AsyncQueue` to `ConcurrentAsyncQueue`

Reply:
you might be right, but there are stronger requirements for concurrent collections and i haven't checked if this implementation satisfies them